### ISSUE TITLE
Create User with AppMetadata

### ISSUE THREAD

**Issue by mosnicholas** [supabase/gotrue#1280]
App metadata seems to be the recommended way to pass app-specific information to the user create flow. If it is specified in the createUser api call, it should be used in the create user GoTrue function.

I'm trying to pass the following flag to `app_metadata: { is_invited: true }` to be able to use in a Postgres trigger and use different logic based on whether a user was invited or not.

I'll be using user_metadata for now, but feels like app_metadata is more appropriate?

```
/**
   * A custom data object to store the user's application specific metadata. This maps to the `auth.users.app_metadata` column.
   *
   * Only a service role can modify.
   *
   * The `app_metadata` should be a JSON object that includes app-specific info, such as identity providers, roles, and other
   * access control information.
   */
  app_metadata?: object
```

https://github.com/supabase/gotrue/blob/40aed622f24066b2718e4509a001026fe7d4b76d/internal/api/admin.go#L352-L362

**Comment by J0** [comment]
Hey @mosnicholas,

Yes, It is possible to create a user and set `app_metadata` as you wish when creating a user.

```
const { data, error } = await supabase.auth.admin.createUser({
  email: 'user@email.com',
  password: 'password',
  app_metadata: { is_invited: 'true' }
})
```

You can choose between using `user_metadata` and `app_metadata` as needed. Please let us know if you run into any issues while using the flow or if there are any doubt/concerns that we haven't  addressed.

Let us know!
Joel

**Comment by mosnicholas** [comment]
@J0 -- I am using a Postgres trigger to check whether users are allowed to sign up / sign in. If a user is invited, I want to short circuit that logic, and always allow the user to be created in our db. I'm passing the `is_invited` flag through the API call to ensure that my postgres trigger can act on the different logic paths based on whether a user was invited or is signing up from the app. 

My understanding from the docs is that the best place to put this flag (`is_invited`) is in the `app_metadata` (ie. app-specific access control info). However, the user is being created in the DB before the `app_metadata` object is passed into the user row, and as a result, I have to use the `user_metadata` object to pass the flag back to the Postgres trigger.

This is why I raised the issue -- essentially, we're creating the app_metadata object as a fixed object, and then updating the user object _after creation_. Feels like maybe we should create the app_metadata object with the values passed to the API on the first pass instead of updating it after the user row is created?

For reference, here's my Postgres trigger:
```
CREATE OR REPLACE FUNCTION public.check_user_email_signup()
 RETURNS trigger
 LANGUAGE plpgsql
 SECURITY DEFINER
AS $function$
DECLARE
  email_domain text;
  domain_exists boolean;
  all_domains text[];
BEGIN
    -- Check if the current user was invited
    IF NEW.raw_user_meta_data->>'is_invited' = 'true' THEN
      return NEW;
    END IF;

    -- If not, check that user is allowed to sign up
    ...

    -- If not, raise an exception
    RAISE EXCEPTION 'Email domain is not allowed';
END;
$function$
;
```

**Comment by J0** [comment]
Hey @mosnicholas,

Thanks for the comprehensive overview - I may be missing something but I agree that we should probably set the `params.AppMetadata` together with the provider fields before creating a user.

At the same time, an issue with changing the behaviour is that people may have update triggers or similar depend on the existing behaviour where the update is performed after the user object is created.

I'll discuss with the team and get back.
Thanks!

**Comment by mosnicholas** [comment]
Hey @J0 any progress on your conversations here? :) I've had to hijack the `raw_user_meta_data` json which has resulted in a bit of data loss :(

<img width="432" alt="Screenshot 2023-11-20 at 3 35 04 PM" src="https://github.com/supabase/gotrue/assets/3606778/0f0641b6-7c8c-4145-b85a-c1184e82b6af">

**Comment by mosnicholas** [comment]
Any progress @J0 ? :)

**Comment by mosnicholas** [comment]
@J0 this set up is causing a separate problem all together. Because app_metadata is updated _after_ creating a user row in the database, we can't listen to the user insert call to update app metadata. Eg. if you have a trigger like so:

```CREATE TRIGGER add_user_tenant_id AFTER INSERT ON auth.users FOR EACH ROW EXECUTE FUNCTION add_custom_claim();```

and that function updates the app_metadata field. There's now a race condition because after the user is created, my trigger function is called, and app metadata is updated separately.

This trigger would be helpful to set up multitenant mapping and ensure that we have the right tenant ids set up properly. Curious what you think!

**Comment by GaryAustin1** [comment]
Without digging into this in detail, it seems that if the update gotrue does on raw_app_meta_data just merged in existing keys from the start then it would not care if the extra metadata was added on the insert trigger or the update trigger.

**Comment by craxrev** [comment]
I have a similar situation where I was following a guide on how to integrate [picket](https://docs.picketapi.com/picket-docs/reference/integrations/supabase#trigger-inserts-on-new-user-creation), seems to be not working so far, the issue is when I add custom `app_metadata` and `user_metadata` using `supabase.auth.admin.createdUser()` method from the sdk it does not get reflected when inserting in a "profiles" table through an `AFTER INSERT` trigger, as in both `app_metadata` and `user_metadata` have default values and not the ones I provided, so I had to also add an `AFTER UPDATE` trigger, here's how I did it:
```sql
-- inserts a row into public.profiles
CREATE OR REPLACE function public.handle_new_user()
RETURNS trigger
LANGUAGE plpgsql
SECURITY DEFINER set search_path = public
AS $$
BEGIN
  IF (TG_OP = 'INSERT') THEN
    INSERT INTO public.profiles (user_id, wallet_address, email, username)
    VALUES (NEW.id, NEW.raw_app_meta_data->>'walletAddress', NEW.email, NEW.raw_user_meta_data->>'username');
  END IF;
  IF (TG_OP = 'UPDATE') THEN
    UPDATE public.profiles
    SET (wallet_address, username) = (NEW.raw_app_meta_data->>'walletAddress', NEW.raw_user_meta_data->>'username')
    WHERE user_id = NEW.id;
  END IF;
  RETURN NEW;
END;
$$;

-- trigger the function every time a user is created
CREATE OR REPLACE trigger on_auth_user_created
  AFTER INSERT ON auth.users
  FOR EACH ROW EXECUTE PROCEDURE public.handle_new_user();

-- trigger the function every time a app/user metadata is updated
CREATE OR REPLACE trigger on_auth_user_metadata_updated
  AFTER UPDATE ON auth.users
  FOR EACH ROW
  WHEN (OLD.raw_app_meta_data <> NEW.raw_app_meta_data OR OLD.raw_user_meta_data <> NEW.raw_user_meta_data)
  EXECUTE PROCEDURE public.handle_new_user()

```

But still not sure if this will cause any problems down the line..

**Comment by activenode** [comment]
Wanted to push this one. For someone implementing triggers based on app_metadata, this isn't possible right now. I also had to write a weird workaround as not even the `AFTER UPDATE` worked as it STILL didn't have the app_metadata - so I wrote an SQL logic that runs on update but also checks if the values are set or not - if not, it just continues. That's kinda weird because it allows for orphans.

**Comment by J0** [comment]
Hey @mosnicholas,

Sorry, I missed the past messages. I've forgotten a large part of the context but you might be able to use [custom claims auth hooks](https://supabase.com/docs/guides/auth/auth-hooks) to achieve your use case.  With respect to this particular issue, I'll check again if it's possible to create the user with app_metadata directly instead of creating and updating.

Hope to circle back with updates soon

**Comment by activenode** [comment]
> With respect to this particular issue, I'll check again if it's possible to create the user with app_metadata directly instead of creating and updating.

That'd be helpful I think as it would allow to not wait for 2 Updates (I had to bypass an INSERT and another UPDATE and then the following UPDATE did contain the data). Thanks!

**Comment by J0** [comment]
Short update here - we're doing a quick scan for update triggers to see how we can move ahead

**Comment by J0** [comment]
Notes for Testing

1.  After this change an update trigger on auth.user should continue to receive the same app_metadata
2. After this change any app_metadata passed in via createUser should also be passed into the payload of an insert trigger on auth.users.

**Comment by J0** [comment]
I'm not sure that 1. will continue to hold. It looks like if we do the update prior to creation:

```
if params.AppMetaData != nil {
	if terr := user.UpdateAppMetaDataWithoutCommit(params.AppMetaData); terr != nil {
 		return terr
	}
	}
    ...
	err = db.Transaction(func(tx *storage.Connection) error {
```

The update doesn't seem to fire on my end.  The overarching concern is that it might affect someone who is using the `app_metadata` field in the update trigger.

**Comment by activenode** [comment]
Thanks for the update. Here's one thought of mine. When I talked to @hf about hooks in general, we were discussing maybe not having so much of triggers either in `auth` schema or in `storage` and going more of the `hook` route. 

And since i mentioned `storage` we might need to have more generalized `Database | Hooks` (UI-wise) instead of just `Auth | Hooks`.

What I'm trying to say is:

Yes, it would be helpful to have this resolved to work with triggers but if we just would have generalized hooks to call like `hook.onUserCreated` or even `storage.onFileUploaded` we could abstract some of the complexity of allowing generic triggers in those schemas away.

HOWEVER, and that's a big one to consider: This MUST work in self-hosted/local development or else this is kinda useless as one won't be able to develop locally and this currently doesn't even work with the existing hooks.

<img width="1192" alt="image" src="https://github.com/supabase/auth/assets/9058762/166824ab-97c3-46b8-b811-8d8185d2bfdf">


Hence I'm saying: Maybe general hooks are the solution which can be triggered by whatever Supabase internal pg function but they MUST work locally and currently they don't.

**Comment by J0** [comment]
> This MUST work in self-hosted/local development or else this is kinda useless as one won't be able to develop locally and this currently doesn't even work with the existing hooks.

Thanks for flagging this - I'll look into what's going on there and get back. For now could you try configuring the Hooks via `config.toml` ? It should work there-  feel free to tag me if it doesn't.

**Comment by J0** [comment]
@activenode can I check where you're seeing the above-mentioned screen? Trying to replicate the issue but realized I'm not sure where it's from as we don't expose the page on CLI. 

Are the screenshots from the self-hosted stck on `supabase/supabase`?  Wondering if you could share more on what your Auth config / yml file looks like there if so.

Lmk

**Comment by activenode** [comment]
Sorry, that one was on me. I just opened the respective URL on local.  Mistaken thinking.

**Comment by vyas-meet** [comment]
> I wrote an SQL logic that runs on update but also checks if the values are set or not - if not, it just continues

Hey, can I get an example of this? Just getting started with supabase and I've got a similar use case where I want to pass on app_data on user creation to trigger insertion into other tables. I got it working with user_metadata only to learn that users can edit it, and since I'm assigning a permission role to users I don't want that.

**Comment by activenode** [comment]
I don't have it at hand but I also don't recommend it, it's rather hacky. But basically all I am doing is:

1. I set `app_metadata { checkValue: true, WHATEVER_OTHER_CONFIG_YOU_NEED: 123 }` (this is Supabase client, not SQL)
2. Then, within the SQL (a trigger that runs on update on `auth.users`) I check if `auth.jwt() -> app_metadata -> 'checkValue'` is defined. Because if it is, it means the metadata is ready.

**Comment by sethtjf** [comment]
Is there any update on this? Running into this issue as well and looking for the best way to resolve.

**Comment by activenode** [comment]
> Is there any update on this? Running into this issue as well and looking for the best way to resolve.

As of now I'd assume this is still the best way to do it https://github.com/supabase/auth/issues/1280#issuecomment-2189528590

**Comment by marcelobaez** [comment]
Comenting to keep track on this. My scenario is as follows:
- I have an Admin-only app (with supabase role key) to create users.
- I have a trigger that calls a function to create a profile whenever a new user is created. 

Everything worked perfectly until i needed to support multi-tenancy, meaning that i must set a "tenant_id" column for each table (profiles included), so my first approach was to send the tenant_id via "app_metadata" while creating the user. I was expecting to read app_metadata when the trigger occurs, but it got erased somehow (only original fields like provider were present).

If I send user_metadata in the user create request, it will still be in the trigger, but thats not secure to rely on, due to user_metadata can be changed or accessed by the browser client.

In the meantime, i will set my app claims (like tenant_id) using the user_metadata in the trigger which i know is working. But then in the auth hook i will erase any user_metadata. It sounds a bit overkill but I cant find any other elegant way

**Comment by activenode** [comment]
> Comenting to keep track on this. My scenario is as follows:
> 
>     * I have an Admin-only app (with supabase role key) to create users.
> 
>     * I have a trigger that calls a function to create a profile whenever a new user is created.
> 
> 
> Everything worked perfectly until i needed to support multi-tenancy, meaning that i must set a "tenant_id" column for each table (profiles included), so my first approach was to send the tenant_id via "app_metadata" while creating the user. I was expecting to read app_metadata when the trigger occurs, but it got erased somehow (only original fields like provider were present).
> 
> If I send user_metadata in the user create request, it will still be in the trigger, but thats not secure to rely on, due to user_metadata can be changed or accessed by the browser client.
> 
> In the meantime, i will set my app claims (like tenant_id) using the user_metadata in the trigger which i know is working. But then in the auth hook i will erase any user_metadata. It sounds a bit overkill but I cant find any other elegant way

Allow us to understand what exact problem is not solved in the above responses, can you elaborate? I did provide a workaround to check when the app_metadata is ready to be set as shown here https://github.com/supabase/auth/issues/1280#issuecomment-2189528590 

If you do this, in combination with jsonb_set, there is no need to "erase" existing data.

**Comment by craxrev** [comment]
@Decs42 It was working fine, no problems I've noticed when I was using that code back then.. now you can that I found 2 ways that I'm still not decided yet which is better:
1- create user through supabase and populate the profile table through a trigger
2- using the new anonymous sign-in feature, insert new profile row and assign the connected user the correct id from the profiles table..

TL;DR: here's the full implementation in a [starter project](https://github.com/Locastic/next-web3-hybrid-starter) I've build, the [main branch](https://github.com/Locastic/next-web3-hybrid-starter) has the trigger code, the [auth/supabase-anonymous-signins branch](https://github.com/Locastic/next-web3-hybrid-starter/tree/auth/supabase-anonymous-signins) without trigger..

Do note that I've dropped using Picket in favour of using RainbowKit directly with a custom authentication, also I noticed that I've dropped the update trigger, I think metadata are correctly available in the insert trigger..

**Comment by joshdance** [comment]
I just fought with this for 2 hours. This is extremely frustrating. I can't find anywhere in the documentation that says 'passing in app_metadata in the createUser API call will write it to the auth table, but it won't immediately be available in standard postgres triggers' :) 

I have a `handle_new_user BEFORE INSERT` trigger on the auth table so when a new user is created, I create a public.users row and link them to the correct company.

The `createUser` call would run the insert trigger, which would create the public.users row, but failed to link the user to a company. After extensive debugging I found that the app_metadata I was passing in wasn't available in the trigger. 

Creating the user:

```
  try {
    const { data: authData, error: createError } = await supabase.auth.admin.createUser({
      email: email,
      password: tempPassword,
      email_confirm: false,

      // PROBLEM DEMONSTRATION: Passing companyId in app_metadata
      // In testing, the 'invited_by_company' value sent here was
      // not available via new.raw_app_meta_data inside a
      // BEFORE INSERT trigger on auth.users.
      app_metadata: {
        invited_by_company: companyId, // e.g., companyId = 1 (a bigint/number)
        is_invited_user: true,
        // other app-specific metadata...
      },

      user_metadata: {
        full_name: name,
      }
    });
```

In the trigger I discovered after setting up a debug log table and writing rows to it, that the app_metadata only contained the provider info:

```
{
  "provider": "email",
  "providers": [
    "email"
  ]
}
```

But when I would check the row in the auth table, it did contain all the app_metadata I had tried to pass in:

```
{
  "provider": "email",
  "providers": [
    "email"
  ],
  "is_invited_user": true,
  "invited_by_company": 1
}
```

This is the trigger function that would would create the new user, but didn't have access to the app_metadata so it would fail to link the user to the company. 

```
CREATE OR REPLACE FUNCTION public.handle_new_user()
RETURNS trigger
LANGUAGE plpgsql
SECURITY DEFINER
SET search_path = public
AS $function$
DECLARE
  user_id bigint;
  company_id_to_link bigint;
BEGIN
  -- PROBLEM DEMONSTRATION:
  -- Attempt to extract 'invited_by_company' from app_metadata.
  -- In testing, this resulted in company_id_to_link being NULL,
  -- even when the API call included it in app_metadata.
  -- The value was unavailable to the BEFORE INSERT trigger at execution time
  -- But was set correctly in the DB
  company_id_to_link := (new.raw_app_meta_data ->> 'invited_by_company')::bigint;

  -- Upsert logic for public.users table:
  -- Insert the new user if email doesn't exist,
  -- otherwise update the auth_id for the existing email.
  INSERT INTO public.users (auth_id, email)
  VALUES (new.id, new.email)
  ON CONFLICT (email) DO UPDATE
    SET auth_id = new.id
  RETURNING id INTO user_id;

  -- Link user to company if company_id was successfully extracted.
  -- This block was skipped because company_id_to_link remained NULL
  -- due to the issue reading from new.raw_app_meta_data.
  IF company_id_to_link IS NOT NULL AND user_id IS NOT NULL THEN
    INSERT INTO public.company_users (user_id, company_id)
    VALUES (user_id, company_id_to_link)
    ON CONFLICT (user_id, company_id) DO NOTHING; -- Optional: handle pre-existing links
  END IF;

  -- Return the new auth.users record to allow the original INSERT to proceed
  RETURN new;
END;
$function$;
```

So what I ended up having to do was switch to using the user_metadata which is available in the insert trigger for some reason:

```
CREATE OR REPLACE FUNCTION public.handle_new_user()
RETURNS trigger
LANGUAGE plpgsql
SECURITY DEFINER
SET search_path = public
AS $function$
DECLARE
  user_id bigint;
  company_id_to_link bigint;
BEGIN
  -- Attempt to extract company_id from user metadata and cast it
  -- Let this fail if the metadata is missing or format is invalid
  company_id_to_link := (new.raw_user_meta_data ->> 'invited_by_company')::bigint;

  -- Check if the email already exists in the public users table
  -- Upsert logic: Insert if new, or update auth_id if email exists
  INSERT INTO public.users (auth_id, email)
  VALUES (new.id, new.email)
  ON CONFLICT (email) DO UPDATE
    SET auth_id = new.id
  RETURNING id INTO user_id;

  -- If a company_id was successfully extracted and a user_id was obtained,
  -- link the user to the company in the company_users table.
  IF company_id_to_link IS NOT NULL AND user_id IS NOT NULL THEN
    INSERT INTO public.company_users (user_id, company_id)
    VALUES (user_id, company_id_to_link)
    -- Handle potential conflict if the link already exists (optional)
    ON CONFLICT (user_id, company_id) DO NOTHING;
  END IF;

  -- Return the new auth.users record to complete the INSERT operation
  RETURN new;
END;
$function$;
```

I believe using user_metadata in this way is secure because I am not using the user_metadata to control access to the company, merely to set the user to the correct company in the DB during the user creation. I think that since the new user can't even access their account until the creation process is finished, they can't login in the split second between when the auth user is created, and when the trigger fires to edit their user_metadata to a company they should not have access to. 

Regardless, this is extremely frustrating and should be documented or fixed. 

The current work arounds as I understand are:

1. Use a trigger on the auth table to create the public.users/profile row and use user_metadata to set the correct user data, but don't use user_metadata to actually enforce access / RLS. 
2. create an UPDATE TRIGGER on the Auth table, which looks at all updates to see if 'checkValue' is defined and skips the work if checkValue is not defined, and when app_metadata eventually gets added to the auth users row, the update trigger will fire again, and since 'checkValue' is defined, it does the trigger. 

Are those accurate? Did I miss any?
